linera-base/src/chrome_trace_exporter.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Chrome trace JSON exporter for OpenTelemetry spans.
+//!
+//! This module provides functionality to export OpenTelemetry spans to Chrome's
+//! trace event format, which can be visualized in `chrome://tracing` or Perfetto UI.
+
+use std::{
+    fs::OpenOptions,
+    io::Write,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+
+use opentelemetry_sdk::{
+    error::OTelSdkError,
+    trace::{SpanData, SpanExporter},
+};
+use serde::{Deserialize, Serialize};
+
+/// Chrome Trace Event Format event.
+///
+/// Follows the Trace Event Format specification:
+/// <https://chromium.googlesource.com/catapult/+/HEAD/docs/trace-event-format.md>
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ChromeTraceEvent {
+    name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    cat: Option<String>,
+    ph: String,
+    ts: u64,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    dur: Option<u64>,
+    pid: u32,
+    tid: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    id: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    args: Option<serde_json::Value>,
+}
+
+/// A collection of Chrome trace events.
+///
+/// This struct wraps a vector of trace events and can be serialized to the
+/// Chrome Trace Event Format JSON structure.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ChromeTrace {
+    trace_events: Vec<ChromeTraceEvent>,
+}
+
+impl ChromeTrace {
+    /// Creates a new empty Chrome trace.
+    pub fn new() -> Self {
+        Self {
+            trace_events: Vec::new(),
+        }
+    }
+
+    /// Adds a trace event to this trace.
+    pub fn add_event(&mut self, event: ChromeTraceEvent) {
+        self.trace_events.push(event);
+    }
+
+    /// Writes this trace to a JSON file at the specified path.
+    pub fn write_to_file(&self, path: &PathBuf) -> std::io::Result<()> {
+        let file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path)?;
+
+        serde_json::to_writer_pretty(file, self)?;
+        Ok(())
+    }
+}
+
+impl Default for ChromeTrace {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// OpenTelemetry span exporter that writes spans to a Chrome trace JSON file.
+///
+/// The exported file can be opened in `chrome://tracing` or the Perfetto UI
+/// for visualization and analysis of distributed traces.
+pub struct ChromeTraceExporter {
+    trace_file: Arc<Mutex<ChromeTraceFile>>,
```

Comment:
If you use `ChromeTrace` here instead, you will not need to manually construct the output JSON file (as you do now down below).